### PR TITLE
Add RFC7919 parameters to DH

### PR DESCRIPTION
**Pull request by snhenson**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

This PR adds RFC7919 support to DH. It adds the parameters from RFC7919, allows them to be selected in a way similar to EC parameters and includes documentation and tests. 

A ctrl is added to pad the DH shared secret with zeroes if necessary: this is needed to support TLS 1.3 DH (TLS 1.2 and earlier does not pad).

The TLS side of this (which will add FFDHE support for TLS 1.3) is *not* included. That will be a separate PR.

### CODE REVIEW

**Comment by snhenson** [comment]
Update: added an optimisation to support constant DH parameters so they don't get duped all the time.

File: crypto/bn/bn_dh.c [hunk 1]
```diff
@@ -104,6 +104,146 @@ static const BN_ULONG dh2048_256_q[] = {
     0x8CF83642A709A097ULL
 };
 
+/* Primes from RFC 7919 */
+static const BN_ULONG ffdhe2048_p[] = {
+    0xFFFFFFFFFFFFFFFFULL, 0x886B423861285C97ULL, 0xC6F34A26C1B2EFFAULL,
```
Comment (dot-asm) [inline crypto/bn/bn_dh.c]: It should be noted that standard says that type of constant will be automatically chosen to accommodate its value. Or in other words constants in question will be of right type without suffixes such as LL. On the other hand standard does specify LL, but some compilers might issue warning when it's assigned to long, or just say things about standard compliance, and others are not compliant and fail to recognize it. Last time I faced this dilemma (in blake2b) I reckoned that omitting suffix minimizes the chances of running into warning. In this case chances of running into problem are smaller than in blake2b, because this is compiled only on 64-bit platforms, while blake2b on both 32- and 64-bit platforms. So this means that this is side note, i.e. no action is expected, just a thing to keep in mind...

File: crypto/bn/bn_dh.c [hunk 2]
```diff
@@ -194,6 +334,147 @@ static const BN_ULONG dh2048_256_q[] = {
     0xA709A097, 0x8CF83642
 };
 
+/* Primes from RFC 7919 */
+
+static const BN_ULONG ffdhe2048_p[] = {
+    0xFFFFFFFF, 0xFFFFFFFF, 0x61285C97, 0x886B4238, 0xC1B2EFFA, 0xC6F34A26,
```
Comment (dot-asm) [inline crypto/bn/bn_dh.c]: Why not U suffix here? Well, as just said type will be auto-selected to accommodate value, so that 0xFFFFFFFF is actually unsigned, but there might be an ambiguity of following character. BN is implemented under assumption that BN_ULONG can be 64 bits wide with THIRTY_TWO_BIT defined. There is no such platforms, but the assumption is still made. For reference, it's manifested by scrupulous masking with BN_MASK2. And ambiguity is in how compiler would treat 32-bit constant without U, when it's assigned to 64-bit value. Well, standard doesn't actually leaves ambiguity(*), but it's not insurance against compiler bugs. Or to summarize, wouldn't U suffix be appropriate here?

(*) 0xFFFFFFFF is supposed to be unsigned and no sign extension is supposed to take place)

File: crypto/bn/bn_dh.c [hunk 3]
```diff
@@ -204,8 +485,12 @@ static const BN_ULONG dh2048_256_q[] = {
                        const BIGNUM _bignum_##x = { (BN_ULONG *) x, \
                         OSSL_NELEM(x),\
                         OSSL_NELEM(x),\
-                        0, BN_FLG_STATIC_DATA };
+                        0, BN_FLG_RDONLY };
```
Comment (dot-asm) [inline crypto/bn/bn_dh.c]: What's rationale behind adding RDONLY? Why STATIC_DATA isn't sufficient?

Comment (snhenson) [reply crypto/bn/bn_dh.c]: Have a look at BN_free and BN_clear_free, both can modify the BIGNUM and there could be compatibility issues if that is changed. So added a new flag BN_FLG_RDONLY to indicate it shouldn't be touched at all.

File: crypto/bn/bn_dh.c (hunk 2 above)
Comment (snhenson) [reply crypto/bn/bn_dh.c]: Well in both cases this follows the pattern that was already there for the (now inadvisable) RFC5114 parameters. Presumably if a compiler had thrown out a warning on those we'd know by now. But if you think it would be better if the RFC7919 (and earlier RFC5114) values should be changed I'll bow to your expertise.

File: crypto/bn/bn_dh.c (hunk 1 above)
Comment (kaduk) [reply crypto/bn/bn_dh.c]: On this (side) note about compiler choosing appropriate type, I do sometimes see a warning "this decimal constant is unsigned only in ISO C90" from asn1_time_test.c, but I did not dig into the details.

File: crypto/bn/bn_dh.c (hunk 2 above)
Comment (snhenson) [reply crypto/bn/bn_dh.c]: Note that bn_nist.c which has been about for a _very_ long time uses the same form. So we'd know about it by now if there were any problems.

Comment (dot-asm) [reply crypto/bn/bn_dh.c]: Note that I'm not saying that not having U suffix doesn't work, only wonder if it would make it less error-prone. It's also kind of a side note. Let us view it as possible overhaul for another time... As for it working elsewhere for long time. Well, somebody on platform where it *could* be a problem might never test it, because it falls out of application's scope... But one again, for some other time, maybe...

File: crypto/bn/bn_lib.c [hunk 4]
```diff
@@ -181,7 +181,7 @@ void BN_clear_free(BIGNUM *a)
 {
     int i;
 
-    if (a == NULL)
+    if (a == NULL || a->flags & BN_FLG_RDONLY)
         return;
     bn_check_top(a);
     if (a->d != NULL) {
```
Comment (dot-asm) [inline crypto/bn/bn_lib.c]: Let's consider lines next to this one. It's OPENSSL_cleanse followed by check for STATIC_DATA and conditional free. It's should be noted however that most (if not all) `a->d` vectors in STATIC_DATA structures reside in read-only segment. So that OPENSSL_cleanse is bound to crash with SEGV in such cases. And it does, just try BN_clear_free(BN_value_one()). The fact that we haven't ever observed it ought to mean that nobody calls it with static data. This in turn ought to mean that it would be appropriate to move OPENSSL_cleanse inside the condition. If we move even further, we'll find ourselves making exactly same argument, i.e. since even most (if not all) STATIC_DATA `*a` themselves residing in read-only segment, so that even second OPENSSL_cleanse should be conditional.  But in this case it becomes equivalent to RDONLY. Another reason for skepticism toward introduction of RDONLY is following. There are more cases where STATIC_DATA is meaningful, e.g. bn_expand_internal, so question is shouldn't RDONLY get same treatment? And when you tie all things up, what would be the difference between the two?

Comment (snhenson) [reply crypto/bn/bn_lib.c]: Well I'm happy to remove RDONLY and make BN_clear_free() not cleanse a->d if STATIC_DATA is set and to not cleanse *a if BN_FLG_MALLOCED is not set either. Now that we've made BIGNUM opaque applications can't meaningfully create a static BIGNUM with BN_FLG_MALLOCED clear.

The aim of that commit is so DH_free() doesn't crash when it tries to free up read only BIGNUM structures. It needlessly calls BN_clear_free() on various parameters. Changing it to BN_free() doesn't help though because BN_free() attempts to modify a a BIGNUM further and uses the mysterious BN_FLG_FREE flag which AFAICS has never been used by anything.

**Comment by snhenson** [comment]
Update: rebased and BN_FLG_RDONLY removed.

File: crypto/bn/bn_lib.c [hunk 5]
```diff
@@ -179,20 +179,17 @@ static void bn_free_d(BIGNUM *a)
 
 void BN_clear_free(BIGNUM *a)
 {
-    int i;
-
     if (a == NULL)
         return;
     bn_check_top(a);
```
Comment (dot-asm) [inline crypto/bn/bn_lib.c]: Formally speaking this bn_check_top should be moved inside the if below, because the kind of thing it does can't be performed on STATIC_DATA. Though I fail to see what purpose does it serve, since we are about to free it anyway. Well, there is assert... Otherwise it can as well be omitted... I mean move or omit :-)

Comment (dot-asm) [reply crypto/bn/bn_lib.c]: Same can be said about BN_free. I.e. move or omit bn_check_top. Just in case assert itself is not the problem, there is bn_pollute. On the other hand, one might be able to argue that STATIC_DATA should be perfectly trimmed, i.e. with top being equal to dmax and with topmost limb being non-zero, so that one might appreciate SEGV as indication that it's not right...

File: crypto/dh/dh_ameth.c [hunk 6]
```diff
@@ -374,13 +374,19 @@ static int dh_cmp_parameters(const EVP_PKEY *a, const EVP_PKEY *b)
...
     BIGNUM *a;
-    if (src) {
-        a = BN_dup(src);
-        if (!a)
-            return 0;
-    } else
+
+    /*
+     * If read only copy the pointer so we don't have
+     * to reallocate it.
```
Comment (dot-asm) [inline crypto/dh/dh_ameth.c]: Could you punctuate the commentary, it's hard to parse...

File: util/libcrypto.num [hunk 7]
```diff
@@ -4400,3 +4400,5 @@ EVP_PKEY_meth_remove                    4343	1_1_1	EXIST::FUNCTION:
 OPENSSL_sk_reserve                      4344	1_1_1	EXIST::FUNCTION:
 CRYPTO_atomic_read                      4345	1_1_1	EXIST::FUNCTION:
 CRYPTO_atomic_write                     4346	1_1_1	EXIST::FUNCTION:
+DH_rfc7919_create                       4347	1_1_1	EXIST::FUNCTION:DH
+DH_get_rfc7919_nid                      4348	1_1_1	EXIST::FUNCTION:DH
```
Comment (dot-asm) [inline util/libcrypto.num]: It seems that in other similar cases subroutine that return DH \* are called DH_get_something. In other words question is what is rationale behind calling this one for DH_rfc7919_create, and not for example DH_get_rfc7919_by_nid...

Comment (snhenson) [reply util/libcrypto.num]: Well it is creating a DH structure and populating it with RFC7919 parameters rather than retrieving it from somewhere else and the choice of name reflects that. I don't really mind too much what it's called. The function `EC_KEY_new_by_curve_name` serves a similar purpose so maybe `DH_new_by_rfc7919_nid`?

File: crypto/bn/bn_lib.c (hunk 5 above)
Comment (snhenson) [reply crypto/bn/bn_lib.c]: Yes I wasn't sure what purpose bn_check_top server either. I'd removed both of them.

File: crypto/dh/dh_ameth.c (hunk 6 above)
Comment (snhenson) [reply crypto/dh/dh_ameth.c]: Done.

File: util/libcrypto.num (hunk 7 above)
Comment (dot-asm) [reply util/libcrypto.num]: Effectively I was referring to DH_get_xxx_yyy in dh/dh_rfc5114.c. Actually I don't have real suggestion for alternative, name just appears inconsistent to me. So I suppose the query was rather to encourage to give it an extra consideration. But new_by_nid sounds better...

**Comment by dot-asm** [comment]
BTW, could you add `[extended tests]` to last commit?

File: util/libcrypto.num (hunk 7 above)
Comment (snhenson) [reply util/libcrypto.num]: Yes I'll change it to DH_new_by_nid and DH_get_nid. There's no reason these should be confined to RFC7919 parameters: we may add new NIDs in future.

**Comment by snhenson** [comment]
Update: rebased, renamed and squashed.

**Comment by snhenson** [comment]
Thanks, pushed.
